[wip] fpm binding master and children processes to specific core(s).

devnexen opened the pull request:
rational is there is git snippets and similar floating around about how to achieve this via command line.
So idea is to do it "natively" especially those above mostly focus on linux whereas more platforms are cpu affinity capable. For now at least, focusing on simple cases, one core id or a range.

iluuu1994 commented:
@devnexen Thank you! I'll have a look soon. :slightly_smiling_face:

iluuu1994 commented on sapi/fpm/php-fpm.conf.in:
```diff
@@ -125,6 +125,14 @@
 ;systemd_interval = 10
 
+; Bind the master process to a cpu set.
+; The value can be one cpu id or a range.
```
```suggestion
; The value can be a cpu id, a range, or a list thereof.
```
Not completely sure about this wording.

iluuu1994 commented on sapi/fpm/tests/cpuaffinity.phpt:
```diff
@@ -0,0 +1,44 @@
+--TEST--
```
Should this test be skipped for single-core environments?

iluuu1994 commented on sapi/fpm/tests/cpuaffinity-fail.phpt:
```diff
@@ -0,0 +1,43 @@
+<?php include "skipif.inc";
+
+if (!str_contains(PHP_OS, 'Linux')) {
+    die('skipped supported only on Linux');
```
FreeBSD supports `cpuset_setaffinity`, is there any reason not to test this?

bukka replied:
Yeah why is this limited to Linux only if `cpuset_setaffinity` is already implemented?

iluuu1994 commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,103 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+		char *cpumasksep;
+
+		fconf.min = strtol(token, &cpumasksep, 0);
+		if (errno || fconf.min > cpumax) {
```
Should this guard against negative numbers?

iluuu1994 commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,103 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+		token = php_strtok_r(NULL, ";", &buf);
+	}
+
+fail:
```
Fail probably shouldn't call `fpm_cpuaffinity_set` at all.

iluuu1994 commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,103 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+#if defined(HAVE_SCHED_SETAFFINITY) || defined(HAVE_CPUSET_SETAFFINITY)
+	cpu_set_t cset;
+#endif
+	long min;
```
I find these slightly misleading, as they are only used as temporary values for `fpm_cpuaffinity_add`. I'd prefer if `min`/`max` were parameters to `fpm_cpuaffinity_add` but it's not a big deal.

bukka replied:
yeah agree, probably not much point to define them in struct as it can be just param. Also agree that it's not a big deal though :)

bukka replied:
Ok I will rephrase it. Please can you change this too.

iluuu1994 commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,103 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+			goto fail;
+		}
+		fconf.max = fconf.min;
+		if (*cpumasksep == '-') {
```
I think this allows trailing characters other than `-` like `42foo`.

iluuu1994 reviewed (commented):
I wonder generally whether the main use-case here would be to pin specific workers to specific CPUs, rather than pinning all workers to a range of CPUs. I've never actually used `setaffinity` so I'm not sure if that makes a big difference or if the scheduler is smart enough to avoid these context switches.

bukka commented on sapi/fpm/fpm/fpm_conf.c:
```diff
@@ -98,6 +98,9 @@ static struct ini_value_parser_s ini_fpm_global_options[] = {
 	{ "process.max",                 &fpm_conf_set_integer,         GO(process_max) },
 	{ "process.priority",            &fpm_conf_set_integer,         GO(process_priority) },
+#if HAVE_FPM_CPUAFFINITY
+	{ "process.cpumask",             &fpm_conf_set_string,          GO(process_cpumask) },
```
Don't forget about `fpm_conf_dump`

bukka reviewed (changes requested):
I think the idea makes sense in general. Added some comments. The ones that stand out for me are following 2 issue.

As I mentioned in one of the the comments I don't like the configuration as it seems very cryptic. I think more declarative approach (potentially introducing more options) would be less error prone and easier to understand IMO.

Support for containers (Docker) is necessary as this would likely be reported as a bug later if it doesn't work correctly. IIRC `_SC_NPROCESSORS_ONLN` reports number of processors in the host rather than what the container is limited to so this might need to be addressed if that's the case.

iluuu1994 commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,102 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
 
+#if HAVE_FPM_CPUAFFINITY
+struct fpm_cpuaffinity_conf {
+#if defined(HAVE_SCHED_SETAFFINITY) || defined(HAVE_CPUSET_SETAFFINITY)
```
I think the idea is to extend this with different checks that are platform specific.

bukka replied:
well it should be then updated or if it's not ready for review yet, then it should be probably a draft PR. I'd rather not to merge anything that has got some part with "maybe in the future"... :)

bukka replied:
This comment still stands.

bukka commented:
@iluuu1994 
> I wonder generally whether the main use-case here would be to pin specific workers to specific CPUs, rather than pinning all workers to a range of CPUs

It sets affinity to all workers (children) in the pool. Pools are the way how the workers are differentiated in FPM. So if you need to select only specific ones, then you need multiple pools.

devnexen commented on configure.ac:
```diff
@@ -570,6 +572,7 @@ AC_CHECK_FUNCS(
 alphasort \
 asctime_r \
 chroot \
+cpuset_setaffinity \
```
For now .... but I have ulterior motives ;)

devnexen commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,102 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+{
+	static long cpuid = LONG_MIN;
+	if (cpuid == LONG_MIN) {
+		cpuid = sysconf(_SC_NPROCESSORS_ONLN);
```
that s a good point indeed.

devnexen commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,102 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+#endif
+}
+
+static void fpm_cpuaffinity_destroy(struct fpm_cpuaffinity_conf *c)
+{
+	// some platform/apis require to allocates data on the heap
+}
```
it will be useful for netbsd portage for example

bukka replied:
It would be better to either implement for all platforms. Ideally I would prefer not to have merged anything that needs some further work or is only useful for some potential future work.

devnexen commented:
> 
> It sets affinity to all workers (children) in the pool.
master process and/or workers (if you do not set worker's, it will inherit master's setting).

devnexen commented on sapi/fpm/php-fpm.conf.in:
```diff
@@ -125,6 +125,14 @@
+; Default Value: not set
+; process.cpumask = "4"
+; process.cpumask = "0-3"
+; process.cpumask = "1-4;8-16;32-48"
```
you re not totally wrong even tough I did on purpose to be ressemblant towards Linux's taskset/FreeBSD cpuset's "mindset".

bukka replied:
Still not sure if it's a good idea but if we want to use this, it would be better to call it `process.cpu_list`. Also why did you choose semicolons - taskset uses commas, right?

bukka commented:
> > It sets affinity to all workers (children) in the pool.
> 
> master process and/or workers (if you do not set worker's, it will inherit master's setting).

Ok I see you apply it in conf. That's of course wrong. The pool config should not change master. If you want to impact master, then it should go to global config and then applied on master.

bukka commented on sapi/fpm/www.conf.in:
```diff
@@ -74,6 +74,14 @@ listen = 127.0.0.1:9000
+; Default Value: inherits master's cpu affinity
+; process.cpumask = "8"
+; process.cpumask = "0-3"
+; process.cpumask = "4-7;10-12"
```
If this format is chosen, it needs more description about the format.

devnexen commented:
@bukka ping :)

bukka commented:
@devnexen I just had a look and `CPU_COUNT` certainly looks like a better solution. Might be better to just error if `sched_getaffinity` fails (returns -1) rather than trying to use sysconf though. Otherwise there are still some unresolved comments so please go through them and either fix them or replay so they all get resolved. Also the FreeBSD failure is related. Once it's all green and comments resolved, I will give a try to see if it all works as expected for me too. :wink:

devnexen commented:
> @devnexen I just had a look and `CPU_COUNT` certainly looks like a better solution. Might be better to just error if `sched_getaffinity` fails (returns -1) rather than trying to use sysconf though. Otherwise there are still some unresolved comments so please go through them and either fix them or replay so they all get resolved. Also the FreeBSD failure is related. Once it's all green and comments resolved, I will give a try to see if it all works as expected for me too. 😉

Think I ve addressed your concerns now :-) cheers

bukka commented:
Cool I will take a proper look early in Feb

bukka commented on sapi/fpm/php-fpm.conf.in:
```diff
@@ -125,6 +125,17 @@
+; The value can be one cpu id, a range or a list thereof.
+;
+; Default Value: not set
+; the master process could be bound to a single core
+; process.cpu_list = "4"
+; a range from the minimum cpuid to the max required
+; process.cpu_list = "0-3"
+; or multiple ranges separated by a comma
+; process.cpu_list = "1-4,8-16,32-48"
```
It would be better to have the format docs before the value rather than actual example. See other entries to get some inspiration. For example https://github.com/php/php-src/blob/64b8306fd8c2252d31481f43eaf1e5874f2bd5b0/sapi/fpm/www.conf.in#L276-L283 .

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -392,6 +402,105 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+
+static void fpm_cpuaffinity_init(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_FPM_CPUAFFINITY)
```
What's the point of this actually. Isn't this already inside this block: https://github.com/php/php-src/pull/10075/files#diff-6b26b2bd2a798d905b4419198f73f61fac2ecc265b4d361d204f565fd91d407fR405

bukka replied:
Yeah so it needs to get removed...

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -392,6 +402,105 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+
+static void fpm_cpuaffinity_add(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_FPM_CPUAFFINITY)
```
What's the point of this actually. Isn't this already inside this block: https://github.com/php/php-src/pull/10075/files#diff-6b26b2bd2a798d905b4419198f73f61fac2ecc265b4d361d204f565fd91d407fR405

bukka replied:
This still stands...

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,103 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+
+	for (i = c->min; i <= c->max; i ++) {
+		if (!CPU_ISSET(i, &c->cset)) {
+			CPU_SET(i, &c->cset);
```
I think this is fine. It would run only in config phase so we don't need some micro optimizations here for undocumented logic.

bukka commented on sapi/fpm/www.conf.in:
```diff
@@ -80,6 +80,17 @@ listen = 127.0.0.1:9000
+; The value can be one cpu id, a range or a list thereof.
+;
+; Default Value: inherits master's cpu affinity
+; the pool process could be bound to a single core
+; process.cpu_list = "8"
+; a range from the minimum cpuid to the max required
+; process.cpu_list = "0-3"
+; or multiple ranges separated by a comma
+; process.cpu_list = "4-7,10-12"
```
Same as above comment for global config - better description would be useful...

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -392,6 +402,105 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+		}
+		fconf.max = fconf.min;
+		if (*cpu_listsep == '-') {
+			if (strlen(cpu_listsep) > 1) {
+				char *err;
+				fconf.max = strtol(cpu_listsep + 1, &err, 0);
+				if (errno || *err != '\0' || fconf.max < fconf.min || fconf.max > cpumax) {
+					return -1;
+				}
```
This is currently not covered by any test in the same way as the list - there should be some tests that would run if more cpus available (doesn't matter that they might not run in the pipeline - they will run on my computer at least)

bukka reviewed (changes requested):
It still needs a bit of work

bukka commented on sapi/fpm/www.conf.in:
```diff
@@ -74,6 +74,12 @@ listen = 127.0.0.1:9000
+; The value can be one cpu id or a range.
+;
+; Default Value: inherits master's cpu affinity
+; process.cpumask = 4-7
```
still requires some tests for those ranges as I noted elsewhere so this is still valid comment

bukka commented:
@devnexen There are still bunch of comments. Please comment on or resolve them once you address them. It will be then clearer what needs to be done. I resolved all those that were done and added some other comments.

devnexen commented:
> @devnexen There are still bunch of comments. Please comment on or resolve them once you address them. It will be then clearer what needs to be done. I resolved all those that were done and added some other comments.

I think I have addressed those, unluckily, vm builders fail at the moment but at least you can try locally the multi range part :-)

bukka commented on sapi/fpm/tests/cpuaffinity-range.phpt:
```diff
@@ -0,0 +1,55 @@
+pm.start_servers = 1
+pm.min_spare_servers = 1
+pm.max_spare_servers = 1
+process.cpu_list = 0-1,2-3
```
Could you maybe change it to `0,2-3` as it will make more sense...

bukka commented on sapi/fpm/tests/cpuaffinity-range.phpt:
```diff
@@ -0,0 +1,55 @@
+--SKIPIF--
+<?php include "skipif.inc";
+
+if (!str_contains(PHP_OS, 'Linux') && !str_contains(PHP_OS, 'FreeBSD')) {
+    die('skipped supported only on Linux and FreeBSD');
+}
+
+if (str_contains(PHP_OS, 'Linux')) {
+    $cmd = 'nproc';
+} else {
+    $cmd = 'sysctl hw.ncpus';
+}
+
+$nproc = intval(exec($cmd));
+if ($nproc < 4) {
+    die('skipped supported only on 4 cores or more environments');
+}
```
Can you integrate those independently to the Tester (static methods) as done elsewhere...

devnexen replied:
good point.

bukka commented on sapi/fpm/php-fpm.conf.in:
```diff
@@ -125,6 +125,18 @@
+;
+; Default Value: not set
+; Valid syntaxes are:
+;       process.cpu_list = "cpu id"                 - to bind to a single core
+;       process.cpu_list = "min cpu id-max cpu id"  - to bind to a core range
+                                                      from minimum cpu id to max
+;       process.cpu_list = "[min cpu id-max cpu id],[min cpu id-max cpu id],..."
+                                                    - to bind to multiple ranges
+                                                      separated by a comma
```
This is actually very confusing description and a bit incorrect. Could you maybe use BNF to describe and keep some examples. I didn't mean to replace the examples completely but to add some generic description.

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -416,6 +521,15 @@ int fpm_unix_init_child(struct fpm_worker_pool_s *wp) /* {{{ */
 	}
+#if HAVE_FPM_CPUAFFINITY
+	if (wp->config->process_cpu_list) {
+
```
Why extra nesting here. Just use single condition.

bukka commented on sapi/fpm/www.conf.in:
```diff
@@ -80,6 +80,18 @@ listen = 127.0.0.1:9000
+; The value can be one cpu id, a range or a list thereof.
+;
+; Default Value: inherits master's cpu affinity
+; Valid syntaxes are:
+;       process.cpu_list = "cpu id"                 - to bind to a single core
+;       process.cpu_list = "min cpu id-max cpu id"  - to bind to a core range
+                                                      from minimum cpu id to max
+;       process.cpu_list = "[min cpu id-max cpu id],[min cpu id-max cpu id],..."
+                                                    - to bind to multiple ranges
+                                                      separated by a comma
```
ditto

bukka reviewed (changes requested):
Added more comments

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -349,6 +358,102 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+
+static void fpm_cpuaffinity_init(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_SCHED_SETAFFINITY) || defined(HAVE_CPUSET_SETAFFINITY)
```
I see that I had this suggestion but it's actually complete as the define is not needed and is already covered now...

devnexen commented:
> Added more comments

Thanks @bukka I did some changes let me know if I forgot something.

devnexen commented:
ping @bukka when you have a moment. 🙂

bukka commented:
I will try to find some time for this in May. Already used almost all my FPM time for this month.

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -392,6 +402,93 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+	while (token) {
+		char *cpu_listsep;
+
+		min = strtol(token, &cpu_listsep, 0);
```
this is too forgiving and allows latters in the string. For example `abbbbb123,1`  results in min and max 0 for the first element and 1 for min max second element. Obviously this should be error. I converted this to a small program:
```c
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

char *php_strtok_r(char *s, const char *delim, char **last)
{
    char *spanp;
    int c, sc;
    char *tok;

    if (s == NULL && (s = *last) == NULL) {
		return NULL;
    }

    /*
     * Skip (span) leading delimiters (s += strspn(s, delim), sort of).
     */
cont:
    c = *s++;
    for (spanp = (char *)delim; (sc = *spanp++) != 0; ) {
		if (c == sc) {
			goto cont;
		}
    }

    if (c == 0)		/* no non-delimiter characters */ {
		*last = NULL;
		return NULL;
    }
    tok = s - 1;

    /*
     * Scan token (scan for delimiters: s += strcspn(s, delim), sort of).
     * Note that delim must have one NUL; we stop if we see that, too.
     */
    for (;;)
    {
		c = *s++;
		spanp = (char *)delim;
		do {
			if ((sc = *spanp++) == c) {
				if (c == 0) {
					s = NULL;
				} else {
					char *w = s - 1;
					*w = '\0';
				}
				*last = s;
				return tok;
			}
		}
		while (sc != 0);
    }
    /* NOTREACHED */
}

int parse(const char *cpu_list_const, int cpumax)
{
    char *token, *buf;
	int min, max;

	int rc = 0;

    char *cpu_list = strdup(cpu_list_const);

    printf("\nparsing %s\n", cpu_list);

	token = php_strtok_r(cpu_list, ",", &buf);

	while (token) {
		char *cpu_listsep;

		min = strtol(token, &cpu_listsep, 0);
		if (errno || min < 0 || min > cpumax) {
			rc = -1;
		}
		max = min;
		if (*cpu_listsep == '-') {
			if (strlen(cpu_listsep) > 1) {
				char *err;
				max = strtol(cpu_listsep + 1, &err, 0);
				if (errno || *err != '\0' || max < min || max > cpumax) {
                    rc = -1;
                    goto end;
				}
			} else {
				rc = -1;
                goto end;
			}
		}

		printf("min: %d, max: %d\n", min, max);

		token = php_strtok_r(NULL, ",", &buf);

    }

end:
    printf("result: %d, cpu list: %s\n", rc, cpu_list);
    free(cpu_list);

	return rc;

}

int main()
{
    parse("1,2,5", 10);
    parse("1-2", 4);
    parse("1-2,4-5,7-10", 12);
    // error cases
    parse("1-8", 4);
    parse("abbbbb123,1", 10);
    parse("123afafkalf,1ffff", 10);
}
```

which results to 
```
parsing 1,2,5
min: 1, max: 1
min: 2, max: 2
min: 5, max: 5
result: 0, cpu list: 1

parsing 1-2
min: 1, max: 2
result: 0, cpu list: 1-2

parsing 1-2,4-5,7-10
min: 1, max: 2
min: 4, max: 5
min: 7, max: 10
result: 0, cpu list: 1-2

parsing 1-8
result: -1, cpu list: 1-8

parsing abbbbb123,1
min: 0, max: 0
min: 1, max: 1
result: 0, cpu list: abbbbb123

parsing 123afafkalf,1ffff
min: 123, max: 123
min: 1, max: 1
result: -1, cpu list: 123afafkalf
```

This needs to be fixed.

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -392,6 +402,93 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
+	cpumax = fpm_cpumax();
+
+	fpm_cpuaffinity_init(&c);
+	token = php_strtok_r(cpu_list, ",", &buf);
```
The problem here is that `php_strtok_r` modifies the value so you are changing the configured value in the config which should not be done. You should copy the list to a new buffer (either on heap allocated or using alloca) before calling this.

krakjoe commented:
This is potentially interesting work.

@devnexen if you don't intend to carry this work on could you close the PR, if you would still like to push this forward, please resolve the conflicts and ping @bukka for review

Assigned @bukka for book-keeping; if this is closed maybe you could add this to your list of one million things to do for fpm ... what's another thing, right ? :)

devnexen commented:
I believe it is already part of his book keeping.

Edit: there [it is](https://github.com/bukka/php-todo/blob/master/php-fpm.md).
